libtenzir/builtins/operators/select.cpp

```diff
@@ -77,10 +77,29 @@ class select_operator final
     return "select";
   }
 
-  auto optimize(expression const& filter, event_order order) const
+  auto optimize(expression const& filter, event_order order,
+                select_optimization const& selection) const
     -> optimize_result override {
-    (void)filter;
-    return optimize_result::order_invariant(*this, order);
+    if (config_.fields.empty() && selection.fields_of_interest.empty()) {
+      return optimize_result::order_invariant(*this, order);
+    }
+    if (!selection.fields_of_interest.empty()) {
+      auto config_fields = config_.fields;
+      auto selection_fields = selection.fields_of_interest;
+      std::sort(config_fields.begin(), config_fields.end());
+      std::sort(selection_fields.begin(), selection_fields.end());
+      std::vector<std::string> intersection;
+      std::set_intersection(config_fields.begin(), config_fields.end(),
+                            selection_fields.begin(), selection_fields.end(),
+                            std::back_inserter(intersection));
+      if (intersection.empty()) {
+        return optimize_result{filter, order, copy(), std::nullopt};
+      }
+      return optimize_result{filter, order, nullptr,
+                             select_optimization(intersection)};
```

Comment:
It seems like this assumes that we are free to reorder fields as part of the optimization, right? I would need to be convinced that this is a good choice, as field order is very visible to the user when looking at the output. The more conservative choice would be to require that the optimization does not change the order of fields.

Reply:
`event_order` is about the order of _events,_ not _fields_. I was asking about the field order. IIUC, your code does not guarantee that the field order is the same before and after optimization.

Reply:
Ya, I see the issue. We want to keep the ordering of the fields the same:
ie `select foo bar baz | select bar foo baz` should be `foo baz` and not `baz foo`

Reply:
That's right. And `select a, b | select b, a` needs to be the same as `select b, a`.

Reply:
Actually, I spoke to @dominiklohmann. We want to proceed under the assumption that `select` reorders (even though it does not atm), to make this future proof.